### ISSUE TITLE
`Edit cells > Transform > Language` support for R

### ISSUE THREAD

**Issue by Henri-Lo** [OpenRefine/OpenRefine#1226]
This is a feature-request to add R support in `Edit cells > Transform > Language`.

**Comment by ettorerizza** [comment]
Would be a great extension of course, but a lot of work. Maybe with [JRI ](https://www.rforge.net/JRI/)?

**Comment by thadguidry** [comment]
Better I think for us long term would be Renjin http://docs.renjin.org/en/latest/introduction.html
I also really like the fact that it uses javax.scripting interfaces http://docs.renjin.org/en/latest/library/evaluating.html
I think just having R within OpenRefine would REALLY expand our user base.

**Comment by akbertram** [comment]
I see there are a few language implementation now - can anyone point me to how these are defined? Would the jython extension be a good model?

**Comment by wetneb** [comment]
@akbertram yes, the jython extension is a good model, as it isolates very well the support for that language.

**Comment by ettorerizza** [comment]
@thadguidry +1000 

R has a very active community, but a bit apart from other programmers. It's a booming language, as evidenced by each StackOverflow survey. R has also a steep learning curve, and being able to use it in a visual interface like that of OR would facilitate its learning.

Most importantly, a full support of R would bring a lot of potentialities to OpenRefine. The Jython extension was a great idea. The problem is that Jython only supports some of the Python modules (those that are not written in C). We can not use a whole range of great packages like numpy, pandas, NLTK and so on. With full support for R (including packages written in C++), there will be not much you cannot do in OpenRefine.

**Comment by akbertram** [comment]
@ettorerizza N.B. Renjin tries to provide the best of both worlds - platform independence and full support for modules with C, C++ and Fortran code. We have a tool chain that compiles these languages to JVM bytecode so they can be used as normal JVM libraries without having to help your users navigate the setup of a Fortran compiler. Builds of the CRAN packages are published to http://packages.renjin.org

**Comment by thadguidry** [comment]
UPDATE: So I posted a nice email to the R Lang users group to let them know we'd love to collaborate and explore some use cases, and not just this one.  Who knows, maybe later after our UI Refresh, they could help build some cool extensions for data exploration.  So we'll see some folks coming into this issue and also asking about things on our mailing list.  Expect it to get busy around here !  Like Alex coming into the picture to take a look at things and see where he can help out with Renjin ! Thanks Alex !!!

**Comment by psychemedia** [comment]
This makes me wonder: if OpenRefine was a Jupyter client, and supported a Jupyter kernel connection on the back of `Edit cells > Transform > Language`, you would be opening up support for all manner of things, especially if there was (Apache Arrow](https://github.com/OpenRefine/OpenRefine/issues/1469) data transfer in place (which would work with or python/pandas dataframes?).

I wonder if there are packages on the back of `https://github.com/minrk/thebelab` that might help with this?

**Comment by thadguidry** [comment]
@psychemedia Do you know R Lang well enough ?  Would you like to have a Hangout this weekend to show me ideas ?

**Comment by magdmartin** [comment]
Linking to the discussion on [OpenRefine user mailing list. ](https://groups.google.com/d/msg/openrefine/3sd302cIplo/jXJnCRECBgAJ)

**Comment by thadguidry** [comment]
So the world has certainly changed since this issue was first opened in just 2 years time.

There is now growing support for Polygot Applications through various tooling.
One of those is [GraalVM](https://https://www.graalvm.org)
Graal can help and do many things actually, [like make Java programs faster or make applications extensible.](https://www.graalvm.org/docs/why-graal/#for-ruby-r-or-python)
Currently GraalVM [has limited support for Ruby, R, Python.](https://www.graalvm.org/docs/why-graal/#for-ruby-r-or-python)
Lots of cool things, like [guest language functions can be eval'ed and used as Java Values](https://www.graalvm.org/docs/graalvm-as-a-platform/embed/#define-guest-language-functions-as-java-values)

I allow others to have full rights on experimentation or ideas :-)

**Comment by wetneb** [comment]
Yes Graal would be a great fit to implement expression languages in OpenRefine. I don't think we can replace our current Jython extension with their Python implementation, as it is still less mature than Jython, but Graal could potentially be used to implement new languages. Also, it should not be too hard to reimplement GREL in Graal, which would probably add a lot of benefits too.

As GREL shows, OpenRefine's expression languages do not need to be fully-fledged programming languages to be useful, so Graal's main shortcoming of the partiality of its implementations is less of a concern than in many other places.

**Comment by thadguidry** [comment]
@wetneb yea exactly m;y feelings for same reasons.

**Comment by wetneb** [comment]
I have very little experience with R but my understanding is that it would not be very useful as an expression language: what is interesting is R is its handling of tables, and that would not really help people write better expressions to compute single values.

Instead, we would need to just let users run an arbitrary R function taking the current table as argument and returning a new table which would replace the current one.
